### ISSUE TITLE
GPIO access as user even with rigth permissions to /sys doesn't work

### ISSUE THREAD

**Issue by olegantonyan** [adafruit/adafruit-beaglebone-io-python#36]
Hi
My problem is described here http://stackoverflow.com/questions/18926979/beaglebone-black-adafruit-io-python-library-gpio-user-permissions

In short: GPIO works only with root even when all permissions to /sys are OK for user and I can control gpio from user's shell, but not with python. 
By experimenting I found that it works if you add delay in event_gpio.c in gpio_export  function after closing file:

```
len = snprintf(str_gpio, sizeof(str_gpio), "%d", gpio);
write(fd, str_gpio, len);
close(fd);
volatile int i = 0; // very cruel delay
for (i = 0; i < 100000000; i++);

// add to list
new_gpio = malloc(sizeof(struct gpio_exp));
```

fsync(fd) and fdatasync(fd) before closing don't help. It seems like some kind of race condition which shows up only for regular user.

**Comment by jstoner** [comment]
Not finding it:

```
[ jstoner@beaglebone:~ ]
3s $> uname -a
Linux beaglebone 3.8.13-bone50 #1 SMP Tue May 13 13:24:52 UTC 2014 armv7l GNU/Linux
3s $> sudo apt-get install linux-image-3.8.13-bone68
[sudo] password for jstoner:
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package linux-image-3.8.13-bone68
E: Couldn't find any package by regex 'linux-image-3.8.13-bone68'
[ jstoner@beaglebone:~ ]
20s $> apt-cache search linux-image
alsa-base - ALSA driver configuration files
linux-headers-3.2.0-4-iop32x - Header files for Linux 3.2.0-4-iop32x
linux-headers-3.2.0-4-ixp4xx - Header files for Linux 3.2.0-4-ixp4xx
linux-headers-3.2.0-4-kirkwood - Header files for Linux 3.2.0-4-kirkwood
linux-headers-3.2.0-4-mv78xx0 - Header files for Linux 3.2.0-4-mv78xx0
linux-headers-3.2.0-4-mx5 - Header files for Linux 3.2.0-4-mx5
linux-headers-3.2.0-4-omap - Header files for Linux 3.2.0-4-omap
linux-headers-3.2.0-4-orion5x - Header files for Linux 3.2.0-4-orion5x
linux-headers-3.2.0-4-versatile - Header files for Linux 3.2.0-4-versatile
linux-headers-3.2.0-4-vexpress - Header files for Linux 3.2.0-4-vexpress
linux-image-3.2.0-4-iop32x - Linux 3.2 for IOP32x
linux-image-3.2.0-4-ixp4xx - Linux 3.2 for IXP4xx
linux-image-3.2.0-4-kirkwood - Linux 3.2 for Marvell Kirkwood
linux-image-3.2.0-4-mv78xx0 - Linux 3.2 for Marvell 78xx0
linux-image-3.2.0-4-mx5 - Linux 3.2 for Freescale i.MX51/53
linux-image-3.2.0-4-omap - Linux 3.2 for TI OMAP3+
linux-image-3.2.0-4-orion5x - Linux 3.2 for Marvell Orion
linux-image-3.2.0-4-versatile - Linux 3.2 for Versatile
linux-image-3.2.0-4-vexpress - Linux 3.2 for ARM Ltd. Versatile Express
linux-image-2.6-iop32x - Linux for IOP32x (dummy package)
linux-image-2.6-ixp4xx - Linux for IXP4xx (dummy package)
linux-image-2.6-kirkwood - Linux for Marvell Kirkwood (dummy package)
linux-image-2.6-mx5 - Linux for Freescale MXC/iMX-based (dummy package)
linux-image-2.6-omap - Linux for TI OMAP3+ (dummy package)
linux-image-2.6-orion5x - Linux for Marvell Orion (dummy package)
linux-image-2.6-versatile - Linux for Versatile (dummy package)
linux-image-iop32x - Linux for IOP32x (meta-package)
linux-image-ixp4xx - Linux for IXP4xx (meta-package)
linux-image-kirkwood - Linux for Marvell Kirkwood (meta-package)
linux-image-mv78xx0 - Linux for Marvell 78xx0 (meta-package)
linux-image-mx5 - Linux for Freescale i.MX51 (meta-package)
linux-image-omap - Linux for TI OMAP3+ (meta-package)
linux-image-orion5x - Linux for Marvell Orion (meta-package)
linux-image-versatile - Linux for Versatile (meta-package)
linux-image-vexpress - Linux for ARM Ltd. Versatile Express (meta-package)
```
my /etc/apt/sources.list:
```
deb http://ftp.us.debian.org/debian/ wheezy main contrib non-free
#deb-src http://ftp.us.debian.org/debian/ wheezy main contrib non-free

deb http://ftp.us.debian.org/debian/ wheezy-updates main contrib non-free
#deb-src http://ftp.us.debian.org/debian/ wheezy-updates main contrib non-free

deb http://security.debian.org/ wheezy/updates main contrib non-free
#deb-src http://security.debian.org/ wheezy/updates main contrib non-free

#deb http://ftp.debian.org/debian wheezy-backports main contrib non-free
##deb-src http://ftp.debian.org/debian wheezy-backports main contrib non-free

deb [arch=armhf] http://debian.beagleboard.org/packages wheezy-bbb main
#deb-src [arch=armhf] http://debian.beagleboard.org/packages wheezy-bbb main
```

I can understand pulling images because of the recent glibc hole, but I'm not seeing anything reasonably current here.

**Comment by RobertCNelson** [comment]
@jstoner You have an older image (/etc/dogtag) my repo and the abilty to install the *.deb as is was added mid August/Sep 2014.  There is no easy way to convert a May 2014 image to this later setup..

Grab a newer snapshot:
http://elinux.org/Beagleboard:BeagleBoneBlack_Debian#Debian_Image_Testing_Snapshots

Regards,

**Comment by jstoner** [comment]
That didn't exactly work. first I had to go back a few steps from that link and comment out that line in /etc/apt/sources.list, and then I installed the archive-keyring, which was at 
```
https://repos.rcn-ee.net/debian/pool/main/r/rcn-ee-archive-keyring/rcn-ee-archive-keyring_2015.01.28~bpo70%2b20150213%2b1_all.deb
```
not at the location on the page (that gave me a 404).

then I added the http://repos.rcn-ee.net/debian lines to /etc/apt/sources.list, and then I did 
```
sudo apt-get install linux-image-3.8.13-bone70 ; sudo reboot
```
and when I rebooted I still saw
```
0s $> uname -a
Linux beaglebone 3.8.13-bone50 #1 SMP Tue May 13 13:24:52 UTC 2014 armv7l GNU/Linux
```

I also tried the image at
```
https://rcn-ee.net/rootfs/bb.org/testing/2015-02-22/console/BBB-eMMC-flasher-debian-7.8-console-armhf-2015-02-22-2gb.img.xz
```
I do want a console-only image, no desktop, but when I burned it to an SD card I couldn't boot a BBB from it. I got the 4 LEDs lit. I don't know if it's worth it to flash it to the BBB. 

I also tried the image at
```
https://rcn-ee.net/rootfs/bb.org/testing/2015-02-22/lxde-4gb/BBB-eMMC-flasher-debian-7.8-lxde-4gb-armhf-2015-02-22-4gb.img.xz
```
and it looks like it boots from the SD, the LEDs blink in the familiar way, but I can't ssh into 192.168.7.2.

**Comment by jstoner** [comment]
and though it does seem to boot, my mac can't seem to mount it, so I can't look in /etc/network/interfaces to see where I should be able to ssh or anything.

**Comment by RobertCNelson** [comment]
@jstoner please be more specific with your comments.  As image snapshots are built weekly with all the fixes from the week. (updates are uploading as I type)

```
https://rcn-ee.net/rootfs/bb.org/testing/2015-03-01/
```

The console vs lxde image does load the usb gadget driver differently on bootup.

Documented right here:
https://github.com/RobertCNelson/boot-scripts/blob/master/boot/am335x_evm.sh#L88

lxde: g_mult = g_serial/g_ether/g_usb-flash
console: g_ether

Note, last week's lxde-4gb ended up being blank, the download was around 5xxKb, it's usually 550Mb-ish..  I fixed/re-uploaded it late friday.  Small goff in my automation script.

Regards,

**Comment by jstoner** [comment]
OK.... more specific. My intentions first:

I am an application developer, trying to get a BBB configured as something more robust than a hobbyist device. I want to use Debian stable. Right now, I want a Debian that is 

* reasonably configured, so I don't have to run my application code as root to be able to speak to devices (which is the requirement that led me here);
* reasonably up to date--including a patch for the recent glibc exposure;
* has no desktop code. I suppose if I have to I can uninstall a bunch of stuff from your lxde image, but I'd prefer not to;
* and ultimately is configured to talk via ethernet, though I'm fine with starting with USB and editing /etc/network/interfaces.

TL;DR of what follows: I put your latest images on a BBB and can't get an ssh session, via USB at 192.168.7.2. If that's the wrong IP what is the right one? 

And do you see anything wrong with the following procedure?

And which image would you start with if you were me?

My procedure:

I am most familiar with the install procedure from 

https://learn.adafruit.com/beaglebone-black-installing-operating-systems/mac-os-x

And I use a similar procedure with your image. Thus:

I begin in my vagrant image, because my mac can't seem to deal with .xz files:
```
vagrant@debian-7:~/shared$ wget https://rcn-ee.net/rootfs/bb.org/testing/2015-03-01/console/BBB-eMMC-flasher-debian-7.8-console-armhf-2015-03-01-2gb.img.xz
--2015-03-02 05:22:36--  https://rcn-ee.net/rootfs/bb.org/testing/2015-03-01/console/BBB-eMMC-flasher-debian-7.8-console-armhf-2015-03-01-2gb.img.xz
Resolving rcn-ee.net (rcn-ee.net)... 69.163.222.213
Connecting to rcn-ee.net (rcn-ee.net)|69.163.222.213|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 51160948 (49M) [text/plain]
Saving to: `BBB-eMMC-flasher-debian-7.8-console-armhf-2015-03-01-2gb.img.xz'

100%[=============================================================================================================================================>] 51,160,948  3.20M/s   in 15s

2015-03-02 05:22:58 (3.17 MB/s) - `BBB-eMMC-flasher-debian-7.8-console-armhf-2015-03-01-2gb.img.xz' saved [51160948/51160948]

vagrant@debian-7:~/shared$ unxz BBB-eMMC-flasher-debian-7.8-console-armhf-2015-03-01-2gb.img.xz
vagrant@debian-7:~/shared$
```
I then switch to my mac host, in the same shared directory. I insert my SD card, and write the image to it:
```
[ jstoner@erzulie:~/shared ]
0s $> sudo dd if=BBB-eMMC-flasher-debian-7.8-console-armhf-2015-03-01-2gb.img of=/dev/disk2 bs=1m
Password:
1700+0 records in
1700+0 records out
1782579200 bytes transferred in 858.016538 secs (2077558 bytes/sec)
```
at the end of that process I get a popup saying 'The disk you inserted was not readable by this computer.' I guess a Mac can't mount an ext4 partition, though I seem to recall doing that before. 

anyway I pull out the SD card, insert it in the BBB, hold down the button by the SD slot and apply power. The LEDs blink a bit and then go into cylon mode for a while and then go dark. 

I remove the SD card from the BBB, remove the power supply, and connect via USB. I wait a moment, it boots, and I try to ssh to the same address the BBB was originally configured with when it came from Adafruit:
```
[ jstoner@erzulie:~/shared ]
880s $> ssh 192.168.7.2
ssh: connect to host 192.168.7.2 port 22: Operation timed out
```
And I got the same result with the lxde image at
```
https://rcn-ee.net/rootfs/bb.org/testing/2015-03-01/lxde-4gb/BBB-eMMC-flasher-debian-7.8-lxde-4gb-armhf-2015-03-01-4gb.img.xz.
```
so do I have the wrong IP?

**Comment by jstoner** [comment]
Aaack--reviewing this thread, it looks like the answer is 'start from source, apply the patch, build the image I want, and then try to use it.'

**Comment by UpolRys** [comment]
Hello Oleg,
First I want to thanks you for your patch, it has really helped me to access GPIO pins as a normal user. I tried your patch on two Beaglebone black boards. On new (zero) beaglebone black (debian, linux 3.8.13-bone-70), I edited the source code of the Adafruit IO library, py_gpio.c as you suggested. And installed adafruit-beaglebone-io-python. Python code I am trying to compile accesses and edits GPIO and I2C pins . gpio_i2c.py file content is 
           1. from Adafruit_I2C import Adafruit_I2C
           2. import Adafruit_BBIO.PWM as PWM
           3. import Adafruit_BBIO.GPIO as GPIO
           4. GPIO.setup("P8_14", GPIO.IN)
           5.GPIO.add_event_detect("P8_14", GPIO.RISING)
           6. i2c = Adafruit_I2C(0x77, 1, False)
           7. i2c.write16(0x01, 0x6852) 
           8. i2c.write16(0x02, 0x0000) 
           
This file runs successfully on new BBB when running as debian user. However, applying similar changes to old BBB giving an runtime error at line 6 when accessing to i2c. This is the error message I am getting 
           Traceback (most recent call last):
            File "integratedFinal.py", line 27, in <module>
            i2c = Adafruit_I2C(0x49, 1, False)
            File "build/bdist.linux-armv7l/egg/Adafruit_I2C.py", line 17, in __init__
            self.bus = smbus.SMBus(busnum if busnum >= 0 else 1)
            IOError: [Errno 13] Permission denied
I guess I did some really improper changes while working on old BBB. I really hope that you can help me to find source of this problem :). I did search on internet with this error message, but no use. I am newbie to BBB and cannot resolve which modification causing this problem (I did so many changes to old BBB).
Thank you for your understanding and help. 
Regards, 
Upol

**Comment by mrandrewandrade** [comment]
I agree on this.We tried the patches and still getting the same error:

    import os
    import sys
    import Adafruit_BBIO.ADC as ADC 

    ADC.setup()

RuntimeError: Unable to setup ADC system. Possible causes are:
    A cape with a conflicting pin mapping is loaded
    A device tree object is loaded that uses the same name for a fragment: helper

**Comment by jonhp** [comment]
And here we are, March 30, 2016 and mrandrewandrade's comment of Nov 18, 2015 is still valid exactly as he wrote it. 
debian 8.3
Linux beaglebone 4.1.15-ti-rt-r43 #1 SMP PREEMPT RT Thu Jan 21 20:13:58 UTC 2016 armv7l GNU/Linux
Will this get fixed ... ever ...?

**Comment by pdp7** [comment]
fyi - I believe that ADC error is because of a difference in the way the ADC works in 4.1 kernel. Adafruit_BBIO requires Linux 3.8 kernel:
[Setting up IO Python Library on BeagleBone Black](https://learn.adafruit.com/setting-up-io-python-library-on-beaglebone-black/overview)

> The most important part here is that you are using a Linux distribution with the 3.8 kernel. This kernel version made some fairly significant changes with how GPIO, PWM and ADC are accessed.

I'd recommend using the Debian 7.9 from here which has the Linux 3.8.13 kernel:
http://beagleboard.org/latest-images

**Comment by jstoner** [comment]
Looking at Debian's site, looks like Wheezy went oldstable 04-25-2015, which means it's over a year out of stable status, and is due to exit support soon. Is there any interest in fixing this? Do you really want to force users onto an unsupported Debian release?

https://wiki.debian.org/DebianWheezy
https://wiki.debian.org/DebianReleases

**Comment by pdp7** [comment]
ADC support for Debian Jessie (Linux kernel 4.1+) has been merged from #102 by @PeteLawler 

I'm noting my progress for PWM support in Debian Jessie in #103

**Comment by yzhou26** [comment]
I have been doing ucontroller programming for a while, but mainly at a user level and BBB is relatively new to me. The work is great but I do recommend in  "PyObject *py_read(PyObject *self, PyObject *args)" function, "value=value/1800" should be changed to "value=value/4096" to make the number sensible and consistent with BBB documentation.

**Comment by razerraz** [comment]
My working version for recent kernel (4.7 - Archlinux), allowing "gpio" group members tu use io.
Unfortunatly, it works only if GPIO.setup() is loaded twice for each handled pin.

## Systemd service : /etc/systemd/system/nonroot-gpio.service
```
[Unit]
Description=Fix system permissions for gpio group

[Service]
Type=simple
ExecStart=/etc/systemd/system/scripts/nonroot-gpio.sh
StandardOutput=journal

[Install]
WantedBy=multi-user.target
```

## Related script : /etc/systemd/system/scripts/nonroot-gpio.sh
```
#!/bin/bash

chown gpio:gpio /sys/class/gpio/unexport /sys/class/gpio/export 
chmod g+w  /sys/class/gpio/unexport /sys/class/gpio/export 

for spidev in /dev/spi*; do
    chown gpio:gpio $spidev
    chmod g+rw $spidev
done
```

## Udev rule : /etc/udev/rules.d/nonroot-gpio.rules
`KERNEL=="gpio*", SUBSYSTEM=="gpio", ACTION=="add", PROGRAM="/etc/udev/scripts/nonroot-gpio.sh %k"`

## Related script : /etc/udev/scripts/nonroot-gpio.sh
```
#!/bin/bash

GPIO_PATH="/sys/class/gpio/$1/"
chown -R gpio:gpio "$GPIO_PATH"
chmod -R g+rw "$GPIO_PATH"
```


Hope it can help some.

**Comment by bubbapizza** [comment]
Yes, there is a weird ass bug in this Adafruit_BBIO module with using pins as a regular user. I am having the exact same issue as the post above:  YOU MUST RUN GPIO.setup() ****TWICE****

Here is how to reproduce it:

**udev rule: /etc/udev/rules.d/10-gpiobug.rules**
 KERNEL=="gpio47", SUBSYSTEM=="gpio", ACTION=="add", PROGRAM="/bin/sh -c 'chown debian /sys%p/value; chown debian /sys%p/direction"

give debian user permissions to export gpio's:
 sudo chgrp debian /sys/class/gpio
 sudo chmod g+rwx /sys/class/gpio 
 sudo chgrp debian /sys/class/gpio/*export
 sudo chmod g+w /sys/class/gpio/*export

**This python program SHOULD work but does NOT**
 #!/usr/bin/python
 #
 # This is a small program to turn on GPIO 47.
 #
 
 import time
 import Adafruit_BBIO.GPIO as GPIO
 
 GPIO47 = "P8_15"
 
 # Initialize pin by calling setup once.
 GPIO.setup(GPIO47, GPIO.OUT)
 
 # Turn on pin.
 GPIO.output(GPIO47, GPIO.HIGH)
 
 time.sleep(1000)


This python program DOES WORK!!!  WHY???
 #!/usr/bin/python
 #
 # This is a small program to turn on GPIO 47.
 #
 
 import time
 import Adafruit_BBIO.GPIO as GPIO
 
 GPIO47 = "P8_15"
 
 # Initialize pin by calling setup twice but wait in between.
 GPIO.setup(GPIO47, GPIO.OUT)
 time.sleep(0.1)
 GPIO.setup(GPIO47, GPIO.OUT)
 
 # Turn on pin.
 GPIO.output(GPIO47, GPIO.HIGH)
 
 time.sleep(1000)

**Comment by nicnicknicky** [comment]
Hello, I recently managed to enable the GPIOs with PWM to work when my ownership is changed to non-root user. This solution was based off reading various threads and I pieced together my own which happened to work.  
This was my [main reference](https://github.com/cnobile2012/RobotControl/tree/master/contrib).
Essentially my contribution was that I went to look at how gpio and pwm were called in the different libraries and implemented permission changes on those files.

My kernel version:
```sh
#uname -a
Linux BeagleBoneBlack 4.4.36-ti-r72 #1 SMP Wed Dec 7 22:29:53 UTC 2016 armv7l GNU/Linux
```
I am still new to quite new to linux so can I verify if I am actually doing the right thing to get it to work?

I have a user *nick* which is part of the group *digital*

Create custom udev rule in `/etc/udev/rules.d/80-gpio-permissions.rules`
 ```
KERNEL=="gpio*", SUBSYSTEM=="gpio", ACTION=="add", PROGRAM="/usr/local/bin/udev-gpio-permissions.sh"
```

Create script in `/usr/local/bin/udev-gpio-permissions.sh`

```sh
  #!/bin/bash
  
  # Result: chown nick:digital /sys/class/gpio
  
  chown -R nick:digital /sys/devices/gpio
  chown -R nick:digital /sys/class/gpio
  chown -R nick:digital /sys/devices/platform/ocp
  chown -R nick:digital /sys/class/pwm
  
  #GPIO ( active_low device direction edge power subsystem uevent value )
  find /sys/class/gpio -type d -exec chmod 2770 {} \;
  find /sys/class/gpio -name "direction" -exec chmod 0660 {} \;
  find /sys/class/gpio -name "edge" -exec chmod 0660 {} \;
  find /sys/class/gpio -name "value" -exec chmod 0660 {} \;
  find /sys/class/gpio -name "active_low" -exec chmod 0660 {} \;
  find /sys/class/gpio -name "device" -exec chmod 0660 {} \;
  find /sys/class/gpio -name "power" -exec chmod 0660 {} \;
  find /sys/class/gpio -name "subsystem" -exec chmod 0660 {} \;
  
  #PWM ( duty_cycle enable period polarity power uevent )
  find /sys/devices/platform/ocp -name "duty_cycle" -exec chmod 0660 {} \;
  find /sys/devices/platform/ocp -name "enable" -exec chmod 0660 {} \;
  find /sys/devices/platform/ocp -name "period" -exec chmod 0660 {} \;
  find /sys/devices/platform/ocp -name "polarity" -exec chmod 0660 {} \;
  find /sys/devices/platform/ocp -name "power" -exec chmod 0660 {} \;
  find /sys/class/pwm -name "duty_cycle" -exec chmod 0660 {} \;
  find /sys/class/pwm -name "enable" -exec chmod 0660 {} \;
  find /sys/class/pwm -name "period" -exec chmod 0660 {} \;
  find /sys/class/pwm -name "polarity" -exec chmod 0660 {} \;
  find /sys/class/pwm -name "power" -exec chmod 0660 {} \;
  
  chmod 0220 /sys/class/gpio/export
  chmod 0220 /sys/class/gpio/unexport
  chmod 0220 /sys/class/pwm/export
  chmod 0220 /sys/class/pwm/unexport
  
  find /sys/class/gpio -name "uevent" -exec chmod 0660 {} \;
  find /sys/class/gpio -name "autosuspend_delay_ms" -exec chmod 0660 {} \;
  find /sys/class/gpio -name "control" -exec chmod 0660 {} \;
  find /sys/class/pwm -name "uevent" -exec chmod 0660 {} \;
  find /sys/devices/platform/ocp -name "uevent" -exec chmod 0660 {} \;
  
  # Additional code for getting AIN pins set up
  ain_activator=/sys/devices/platform/bone_capemgr/slots
  chown -R nick:digital $ain_activator/
  chmod -R 2770 $ain_activator/
  ```

Ensure the above file is executable
  ```
  ls -al /usr/local/bin/udev-gpio-permissions.sh
  -rwxr--r-- 1 root staff 896 Jan 31 11:23 /usr/local/bin/udev-gpio-permissions.sh
  ```
Run `udevadm control --reload-rules`

Reboot and test!
